File: src/shared/chests.ts
/**
 * Mystery chest roll — variable-ratio bonus XP on completions. Fixed +5 XP
 * goes numb fast; an occasional surprise bonus keeps the reward circuit
 * interested. Pure and rand-injectable for tests; RNG runs only in the
 * service worker.
 */

export const CHEST_DROP_RATE = 0.15;

/** Bonus tiers with cumulative-scan weights (must sum to 1) */
const TIERS: readonly { bonusXp: number; weight: number }[] = [
  { bonusXp: 10, weight: 0.7 },
  { bonusXp: 25, weight: 0.25 },
  { bonusXp: 50, weight: 0.05 },
];

/** Returns the chest's bonus XP, or null when no chest drops */
export function rollChest(rand: () => number = Math.random): number | null {
  if (rand() >= CHEST_DROP_RATE) return null;
  let roll = rand();
  for (const tier of TIERS) {
    if (roll < tier.weight) return tier.bonusXp;
    roll -= tier.weight;
  }
  return TIERS[TIERS.length - 1].bonusXp;
}

File: ios/ADHDReaderCore/Sources/ADHDReaderCore/Chests.swift
import Foundation

/// Mystery-chest roll — variable-ratio bonus XP, ported from `src/shared/chests.ts`.
/// `rand` is injectable so tests are deterministic.
public enum Chests {
    public static let dropRate = 0.15

    /// Bonus tiers with cumulative-scan weights (sum to 1).
    static let tiers: [(bonusXp: Int, weight: Double)] = [
        (10, 0.7), (25, 0.25), (50, 0.05),
    ]

    /// Returns the chest's bonus XP, or nil when no chest drops.
    public static func rollChest(_ rand: () -> Double = { Double.random(in: 0..<1) }) -> Int? {
        if rand() >= dropRate { return nil }
        var roll = rand()
        for tier in tiers {
            if roll < tier.weight { return tier.bonusXp }
            roll -= tier.weight
        }
        return tiers.last!.bonusXp
    }
}
